crates/bevy_ecs/src/observer/mod.rs

```diff
@@ -0,0 +1,599 @@
+//! Types for creating and storing [`Observer`]s
+
+mod builder;
+mod entity_observer;
+mod runner;
+
+use std::marker::PhantomData;
+
+pub use builder::*;
+pub(crate) use entity_observer::*;
+pub use runner::*;
+
+use crate::{
+    archetype::ArchetypeFlags,
+    query::DebugCheckedUnwrap,
+    system::{EmitEcsEvent, IntoObserverSystem},
+    world::*,
+};
+
+use bevy_ptr::{Ptr, PtrMut};
+use bevy_utils::{EntityHashMap, HashMap};
+
+use crate::{component::ComponentId, prelude::*, world::DeferredWorld};
+
+/// Type used in callbacks registered for observers.
+pub struct Observer<'w, E, B: Bundle = ()> {
+    data: &'w mut E,
+    trigger: ObserverTrigger,
+    _marker: PhantomData<B>,
+}
+
+impl<'w, E, B: Bundle> Observer<'w, E, B> {
+    pub(crate) fn new(data: &'w mut E, trigger: ObserverTrigger) -> Self {
+        Self {
+            data,
+            trigger,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Returns the event id for the triggering event
+    pub fn event(&self) -> ComponentId {
+        self.trigger.event
+    }
+
+    /// Returns a reference to the data associated with the event that triggered the observer.
+    pub fn data(&self) -> &E {
+        self.data
+    }
+
+    /// Returns a mutable reference to the data associated with the event that triggered the observer.
+    pub fn data_mut(&mut self) -> &mut E {
+        self.data
+    }
+
+    /// Returns a pointer to the data associated with the event that triggered the observer.
+    pub fn data_ptr(&self) -> Ptr {
+        Ptr::from(&self.data)
+    }
+
+    /// Returns the entity that triggered the observer, could be [`Entity::PLACEHOLDER`].
+    pub fn source(&self) -> Entity {
+        self.trigger.source
+    }
+}
+
+#[derive(Default, Clone)]
+pub(crate) struct ObserverDescriptor {
+    events: Vec<ComponentId>,
+    components: Vec<ComponentId>,
+    sources: Vec<Entity>,
+}
+
+/// Metadata for the source triggering an [`Observer`],
```

Comment:
These docs don't provide the context needed by maintainers or users, especially for a public type.